applications/OptimizationApplication/custom_utilities/control/sigmoidal_projection_utils.cpp

```diff
@@ -0,0 +1,290 @@
+            x1 = xLimits[i];
+            x2 = xLimits[i + 1];
+            y1 = yLimits[i];
+            y2 = yLimits[i + 1];
+            break;
+        }
+    }
+
+    if (xValue > xLimits[size - 1]) {
+        x1 = xLimits[size - 2];
+        x2 = xLimits[size - 1];
+        y1 = yLimits[size - 2];
+        y2 = yLimits[size - 1];
+    }
+
+    double pow_val = -2.0 * beta * (xValue - (x1 + x2) / 2)  * penalFac;
+    double limit = std::log1p(std::numeric_limits<double>::max());
+    pow_val = std::clamp(pow_val, -limit, limit);
+    return (y2-y1) * (1.0/std::pow(1+std::exp(pow_val),penalFac+1)) * penalFac * 2.0 * beta * std::exp(pow_val);
+
+}
+
+template<class TContainerType>
+ContainerExpression<TContainerType> SigmoidalProjectionUtils::ProjectForward(
+    ContainerExpression<TContainerType>& rInputExpression,
+    std::vector<double> xValues,
+    std::vector<double> yValues,
+    double beta,
+    int penalFac)
+{
+    KRATOS_TRY
+
+    CheckXYVectors(xValues,yValues);
+
+    const auto& r_input_expression = rInputExpression.GetExpression();
+    const IndexType local_size_1 = rInputExpression.GetItemComponentCount();
+    const IndexType number_of_entities_1 = rInputExpression.GetContainer().size();
+
+    ContainerExpression<TContainerType> output_container(rInputExpression.GetModelPart());
+    auto p_flat_data_expression = LiteralFlatExpression<double>::Create(number_of_entities_1, rInputExpression.GetItemShape());
+    output_container.SetExpression(p_flat_data_expression);
+    auto& r_output_expression = *p_flat_data_expression;
+
+
+    IndexPartition<IndexType>(number_of_entities_1).for_each([&r_input_expression, &r_output_expression, &xValues, &yValues, &beta, &penalFac, local_size_1](const IndexType EntityIndex) {
+            const IndexType local_data_begin_index = EntityIndex * local_size_1;
+            for (IndexType i = 0; i < local_size_1; ++i) {
+                double input_value = r_input_expression.Evaluate(EntityIndex, local_data_begin_index, i);
+                double projected_value = ProjectValueForward(input_value,xValues,yValues,beta,penalFac);
+                r_output_expression.SetData(local_data_begin_index, i, projected_value);
+            }
+        });
+
+    return output_container;
+
+    KRATOS_CATCH("");
+}
+
+template<class TContainerType>
+ContainerExpression<TContainerType> SigmoidalProjectionUtils::ProjectBackward(
+    ContainerExpression<TContainerType>& rInputExpression,
+    std::vector<double> xValues,
+    std::vector<double> yValues,
+    double beta,
+    int penalFac)
+{
+    KRATOS_TRY
+
+    CheckXYVectors(xValues,yValues);
+
+    const auto& r_input_expression = rInputExpression.GetExpression();
+    const IndexType local_size_1 = rInputExpression.GetItemComponentCount();
+    const IndexType number_of_entities_1 = rInputExpression.GetContainer().size();
+
+    ContainerExpression<TContainerType> output_container(rInputExpression.GetModelPart());
```

Comment:
```suggestion
    ContainerExpression<TContainerType> output_container(*rInputExpression.pGetModelPart());
```

Getting `non const ModelPart` from `const rInputExpression`